### PR TITLE
Renamed several types in System.Buffers.Primitives 

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
Renamed:
- Memory\<T\> to Buffer\<T\>
- ReadOnlyMemory\<T\> to ReadOnlyBuffer\<T\>
- MemoryHandle to BufferHandle
- OwnedMemory\<T\> to Memory\<T\>

Moved all the types to System.Buffers namespace
Move BufferPool to System.Buffers

Made the following types internal:
- OwnedArray\<T\>: use OwnedBuffer\<T\> cast and Create method to create buffers over T[] and ArraySegment\<T\>
- ManagedBufferPool: use BufferPool.Default

The names might not be final, but I think they move us closer to where we want to be

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
cc: @ahsonkhan, @davidfowl, @pakrym

File: src/System.Buffers.Primitives/System/Buffers/Memory.cs [hunk 1]
```diff
@@ -97,45 +97,45 @@ public void CopyTo(Span<T> span)
             Span.CopyTo(span);
         }
 
-        public void CopyTo(Memory<T> memory)
+        public void CopyTo(Buffer<T> memory)
```
Comment (shiftylogic) [inline src/System.Buffers.Primitives/System/Buffers/Memory.cs]: Do we care about the name of the arguments, especially for public surface?

File: samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs [hunk 2]
```diff
@@ -25,19 +25,19 @@ static void Free(byte[] array)
...
         public IMemoryList<byte> Rest => _next;
 
         public int WrittenByteCount => _written;
 
-        ReadOnlyMemory<byte> IReadOnlyMemoryList<byte>.First => Memory;
+        ReadOnlyBuffer<byte> IReadOnlyMemoryList<byte>.First => Memory;
 
         IReadOnlyMemoryList<byte> IReadOnlyMemoryList<byte>.Rest => _next;
 
         // TODO: maybe these should be renamed to no conflict with OwnedMemory<T>.Length?
```
Comment (ahsonkhan) [inline samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs]: Does this comment still apply?

File: src/System.Buffers.Experimental/System/Buffers/IReadOnlyMemoryList.cs [hunk 3]
```diff
@@ -5,10 +5,10 @@
 using System.Collections.Sequences;
 
 namespace System.Buffers {
-    public interface IReadOnlyMemoryList<T> : ISequence<ReadOnlyMemory<T>>
+    public interface IReadOnlyMemoryList<T> : ISequence<ReadOnlyBuffer<T>>
```
Comment (ahsonkhan) [inline src/System.Buffers.Experimental/System/Buffers/IReadOnlyMemoryList.cs]: Do we want to rename the interface as well?

File: src/System.Buffers.Experimental/System/Buffers/OwnedNativeMemory.cs [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 namespace System.Buffers
 {
-    public class OwnedNativeMemory : OwnedMemory<byte>
+    public class OwnedNativeMemory : OwnedBuffer<byte>
```
Comment (ahsonkhan) [inline src/System.Buffers.Experimental/System/Buffers/OwnedNativeMemory.cs]: Change OwnedNativeMemory?

File: src/System.Buffers.Primitives/System/Buffers/Memory.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Primitives/System/Buffers/Memory.cs]: Yeah, I will change them

File: src/System.Buffers.Experimental/System/Buffers/IReadOnlyMemoryList.cs (hunk 3 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/System/Buffers/IReadOnlyMemoryList.cs]: Yup. Thanks.

File: src/System.Buffers.Experimental/System/Buffers/OwnedNativeMemory.cs (hunk 4 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/System/Buffers/OwnedNativeMemory.cs]: Will do. thanks.

File: samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs]: I think so, it's conflicting with Length

File: src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 namespace System.Buffers
 {
-    public sealed class OwnedArray<T> : OwnedMemory<T>
+    internal sealed class OwnedArray<T> : OwnedBuffer<T>
```
Comment (davidfowl) [inline src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs]: Why is this internal?

File: src/System.Buffers.Primitives/System/Buffers/OwnedMemory.cs [hunk 6]
```diff
@@ -207,13 +207,18 @@ static IntPtr Add(IntPtr start, int index)
...
-            throw new ObjectDisposedException(nameof(Memory<T>));
+            throw new ObjectDisposedException(nameof(Buffer<T>));
         }
         void ThrowArgHelper()
         {
             throw new ArgumentOutOfRangeException();
         }
         #endregion
+
+        public static OwnedBuffer<T> Create(ArraySegment<T> segment)
```
Comment (davidfowl) [inline src/System.Buffers.Primitives/System/Buffers/OwnedMemory.cs]: Ah

File: src/System.Buffers.Primitives/System/Runtime/DebuggerViews.cs [hunk 7]
```diff
@@ -1,42 +1,43 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Buffers;
 using System.Diagnostics;
 
 namespace System.Runtime
 {
     internal class MemoryDebuggerView<T>
```
Comment (davidfowl) [inline src/System.Buffers.Primitives/System/Runtime/DebuggerViews.cs]: `BufferDebuggerView`

File: src/System.IO.Pipelines/ReadableBuffer.cs [hunk 8]
```diff
@@ -250,14 +250,14 @@ public static ReadableBuffer Create(byte[] data, int offset, int length)
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length);
             }
 
-            var buffer = new OwnedArray<byte>(data);
+            OwnedBuffer<byte> buffer = data;
```
Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: `var var`

Comment (ahsonkhan) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Shouldn't the type be stated explicitly in this case?

**Comment by Drawaes** [comment]
You are walking into a mine field... asp.net and .net team have different guidelines

File: src/System.Buffers.Primitives/System/Runtime/DebuggerViews.cs (hunk 7 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Primitives/System/Runtime/DebuggerViews.cs]: fixed. Thanks

File: src/System.IO.Pipelines/ReadableBuffer.cs (hunk 8 above)
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: This is not about guidelines/conventions. var buffer = data would declare and assign T[] buffer local. We need buffer to be OwnedBuffer\<byte\> not byte[]

File: src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs (hunk 5 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs]: There are no new APIs added to OwnedArray and so no need for the type to be public.

File: src/System.Buffers.Primitives/System/Buffers/OwnedMemory.cs (hunk 6 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Primitives/System/Buffers/OwnedMemory.cs]: :-)

File: src/System.IO.Pipelines/ReadableBuffer.cs (hunk 8 above)
Comment (benaadams) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Implicit ctor via cast via declaration?

Wouldn't `var buffer = (OwnedBuffer<byte>)data;` be better or expose a `Create(T[] array)` function or people might either get confused how to create or start using (due to not working out how to create the buffer)
```
var buffer = Create(new ArraySegment<T>(array, 0, array.Length);
```
Which would be inefficent

**Comment by omariom** [comment]
I like the new names.

**Pull request merged**